internal/proxy/connection_manager.go

```diff
@@ -0,0 +1,190 @@
+package proxy
+
+import (
+	"context"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+
+	"github.com/milvus-io/milvus-proto/go-api/commonpb"
+
+	"github.com/milvus-io/milvus/pkg/log"
+)
+
+const (
+	// we shouldn't check this too frequently.
+	defaultConnCheckDuration  = 2 * time.Minute
+	defaultTTLForInactiveConn = 24 * time.Hour
+)
+
+type connectionManager struct {
+	mu sync.RWMutex
+
+	initOnce sync.Once
+	stopOnce sync.Once
+
+	closeSignal chan struct{}
+	buffer      chan int64
+	duration    time.Duration
+	ttl         time.Duration
+
+	clientInfos map[int64]clientInfo
+}
+
+type connectionManagerOption func(s *connectionManager)
+
+func withDuration(duration time.Duration) connectionManagerOption {
+	return func(s *connectionManager) {
+		s.duration = duration
+	}
+}
+
+func withTTL(ttl time.Duration) connectionManagerOption {
+	return func(s *connectionManager) {
+		s.ttl = ttl
+	}
+}
+
+func (s *connectionManager) apply(opts ...connectionManagerOption) {
+	for _, opt := range opts {
+		opt(s)
+	}
+}
+
+func (s *connectionManager) init() {
+	s.initOnce.Do(func() {
+		go s.checkLoop()
+	})
+}
+
+func (s *connectionManager) stop() {
+	s.stopOnce.Do(func() {
+		close(s.closeSignal)
+	})
+}
+
+func (s *connectionManager) checkLoop() {
+	t := time.NewTicker(s.duration)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-s.closeSignal:
+			log.Info("connection manager closed")
+			return
+		case identifier := <-s.buffer:
+			s.update(identifier)
+		case <-t.C:
+			s.removeLongInactiveClients()
+		}
+	}
+}
+
+func (s *connectionManager) register(ctx context.Context, identifier int64, info *commonpb.ClientInfo) {
+	cli := clientInfo{
+		ClientInfo:     info,
+		identifier:     identifier,
+		lastActiveTime: time.Now(),
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.clientInfos[identifier] = cli
+	cli.ctxLogRegister(ctx)
+}
+
+func (s *connectionManager) keepActive(identifier int64) {
+	// make this asynchronous and then the rpc won't be blocked too long.
+	s.buffer <- identifier
+}
+
+func (s *connectionManager) update(identifier int64) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	cli, ok := s.clientInfos[identifier]
+	if ok {
+		cli.lastActiveTime = time.Now()
+		s.clientInfos[identifier] = cli
+	}
+}
+
+func (s *connectionManager) removeLongInactiveClients() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	candidates := s.findLongInactiveClients()
```

Comment:
just put logic of "findLongInactiveClients" inside this function and remove selected keys while range loop?